fastrand.go

```diff
@@ -42,36 +49,52 @@ func init() {
 
 // Read fills b with random data. It always returns len(b), nil.
 func (r *randReader) Read(b []byte) (int, error) {
-	// Update the progress and use the result + the seed entropy to form a
-	// unique seed to hash.
-	progress := atomic.AddUint64(&r.progress, 1)
-
-	// If it seems like the progress is about to reset, increment the first
-	// entropy bytes. Without this check, the cycle time is 2^64 iterations, but
-	// with the check it is a secure 2^128 iterations.
-	if progress > 1<<63 {
-		atomic.AddUint64((*uint64)(unsafe.Pointer(&r.entropy[0])), 1)
-		atomic.StoreUint64(&r.progress, 0)
+	// Grab a unique counter from the reader, while atomically updating the
+	// counter so that concurrent callers also end up with unique values.
+	counter := atomic.AddUint64(&r.counter, 1)
+	counterExtra := atomic.LoadUint64(&r.counterExtra)
+	// Update the second 64 bits of the counter if the first 64 bits are close
+	// to wrapping around. It is possible that the second 64 bits of the counter
+	// is updates multiple times by several concurrent threads. This wastes part
+	// of the counter space (up to 2^63 items each time), however the overall
+	// space is large enough (2^128) that wasting some every reset does not make
+	// it any more likely that the caller exhaust the whole possible search
+	// space.
+	if counter > 1<<63 {
+		atomic.AddUint64(&r.counterExtra, 1)
+		atomic.StoreUint64(&r.counter, 0)
```

Comment:
ok, I think the solution is actually not so bad.

Increment extra at `counter == 1<<63` and again at `counter == math.MaxUint64`. You have 2^63 time between each call, and it also guarantees that by the time it's wrapped all the way around back to the increment point, the counter is definitely higher than it's ever been before even if the threads are overlapping a bit.